Stop using Vector::unsafeAppendWithoutCapacityCheck() in mathProtoFuncHypot()

cdumez opened the pull request:
#### e3fed8d9129f924bbdc6c6d108d1e7427231dd96
<pre>
Stop using Vector::unsafeAppendWithoutCapacityCheck() in mathProtoFuncHypot()
<a href="https://bugs.webkit.org/show_bug.cgi?id=265150">https://bugs.webkit.org/show_bug.cgi?id=265150</a>

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: <a href="https://commits.webkit.org/270991@main">https://commits.webkit.org/270991@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0410526e46b2dfc7062e0a3aedf1bfe67599cc2b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26997 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5613 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/28233 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/29208 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24685 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/7473 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3005 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/24555 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/27259 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4437 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23177 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3885 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3961 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24175 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29843 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23532 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24641 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24590 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/30166 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/26181 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3973 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/2171 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/28080 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5435 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33634 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6489 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4440 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7280 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/4344 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/JavaScriptCore/runtime/MathObject.cpp:
```diff
@@ -190,13 +190,12 @@ JSC_DEFINE_HOST_FUNCTION(mathProtoFuncHypot, (JSGlobalObject* globalObject, Call
-        RETURN_IF_EXCEPTION(scope, { });
-        args.unsafeAppendWithoutCapacityCheck(argument);
-    }
+        RETURN_IF_EXCEPTION(scope, std::nullopt);
```
The old code used to bail out if an exception happened without doing further work. This code instead continues to evaluate future arguments although it then does return. We should not change that semantic. Evaluating arguments can have side effects. We can make a test that covers this if we like.

cdumez replied:
The new code bails out when an exception happens too. I do not think I changed behavior. This constructor is not a "compactMap"-like constructor:
```
            for (size_t i = 0; i < size; ++i) {
                if (auto item = valueGenerator(i))
                    unsafeAppendWithoutCapacityCheck(WTFMove(*item));
                else
                    return;
            }
```

cdumez replied:
Maybe we should clarify this in some way? Maybe we could add a template parameter to specify whether to skip or abort when getting std::nullopt?

darinadler reviewed (approved):
OK. I was surprised to learn about this.

I wonder if we have test coverage. Don’t need to add it now necessarily.

cdumez committed:
Stop using Vector::unsafeAppendWithoutCapacityCheck() in mathProtoFuncHypot()
https://bugs.webkit.org/show_bug.cgi?id=265150

Reviewed by Darin Adler.

* Source/JavaScriptCore/runtime/MathObject.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):

Canonical link: https://commits.webkit.org/270991@main
